src/backend/catalog/storage_pending_deletes.c

```diff
@@ -9,15 +9,277 @@
+ */
+dsa_pointer
+PdlShmemAdd(const RelFileNodePendingDelete * relnode, TransactionId xid)
+{
+	if (!is_tracking_enabled() || xid == InvalidTransactionId ||
+		MyBackendId == InvalidBackendId)
+		return InvalidDsaPointer;
+
+	PdlAttachDsa();
+
+	PendingDeleteListNode *node;
+	const dsa_pointer node_dsa = dsa_allocate(pendingDeletesDsa,
+											  sizeof(*node));
+
+	node = dsa_get_address(pendingDeletesDsa, node_dsa);
+	*node = (PendingDeleteListNode)
+	{
+		.xrelnode =
+		{
+			.relnode = *relnode,
+			.xid = xid
+		},
+		.prev = InvalidDsaPointer
+	};
+
+	PendingDeletesList *list = &BackendsPendingDeletes->list[MyBackendId];
+
+	LWLockAcquire(list->lock, LW_EXCLUSIVE);
+	node->next = list->head;
+	if (DsaPointerIsValid(node->next))
+	{
+		PendingDeleteListNode *next_node = (PendingDeleteListNode *)
+			dsa_get_address(pendingDeletesDsa, node->next);
+
+		next_node->prev = node_dsa;
+	}
+	list->head = node_dsa;
+	LWLockRelease(list->lock);
+
+	return node_dsa;
+}
+
+/*
+ * Remove pending delete node from the list of current backend.
+ * node_ptr is a ptr to already added node (see PdlShmemAdd)
+ */
+void
+PdlShmemRemove(dsa_pointer node_ptr)
+{
+	if (!is_tracking_enabled() || MyBackendId == InvalidBackendId)
+		return;
+
+	Assert(DsaPointerIsValid(node_ptr));
+
+	PendingDeletesList *list = &BackendsPendingDeletes->list[MyBackendId];
+	const PendingDeleteListNode *node =
+		dsa_get_address(pendingDeletesDsa, node_ptr);
+
+	LWLockAcquire(list->lock, LW_EXCLUSIVE);
+	if (DsaPointerIsValid(node->next))
+	{
+		PendingDeleteListNode *next_node = dsa_get_address(pendingDeletesDsa,
+														   node->next);
+
+		next_node->prev = node->prev;
+	}
+
+	if (DsaPointerIsValid(node->prev))
+	{
+		PendingDeleteListNode *prev_node = dsa_get_address(pendingDeletesDsa,
+														   node->prev);
+
+		prev_node->next = node->next;
+	}
+	else
+		list->head = node->next;
+
+	LWLockRelease(list->lock);
+
+	dsa_free(pendingDeletesDsa, node_ptr);
+}
+
+/*
+ * Collect info about pending deletes from all backends and return
+ * the accumulated result. Return NULL if there are no nodes in the lists.
+ * Note: the returned result is palloc'ed. Caller is responsible for
+ * freeing it.
+ */
+PendingRelXactDeleteArray *
+PdlXLogShmemDump(void)
+{
+	PdlAttachDsa();
+
+	PendingRelXactDeleteArray *ret = NULL;
+
+	Size		size = offsetof(PendingRelXactDeleteArray, array);
+
+	for (int i = 0; i < MaxBackends; i++)
+	{
+		PendingDeletesList *list = &BackendsPendingDeletes->list[i];
+
+		LWLockAcquire(list->lock, LW_SHARED);
+
+		for (dsa_pointer pdl_node_dsa = list->head;
+			 DsaPointerIsValid(pdl_node_dsa);)
+		{
+			const PendingDeleteListNode *pdl_node =
+				dsa_get_address(pendingDeletesDsa, pdl_node_dsa);
+
+			size += sizeof(*ret->array);
+			if (ret != NULL)
+				ret = repalloc(ret, size);
+			else
+			{
+				ret = palloc(size);
+				ret->count = 0;
+			}
+
+			ret->array[ret->count++] = pdl_node->xrelnode;
```

Comment:
Mixing increment with other expression is not a good practice (https://pvs-studio.com/en/docs/warnings/v2554/) . I suggest:
```suggestion
			ret->array[ret->count] = pdl_node->xrelnode;
			ret->count++;
```

Reply:
What is the problem in this case? The link contains an example of undefined behavior, but in this case there is no undefined behavior. It is better to write in one line, because we should fill array element and increment ret->count in one time. The current implementation emphasizes that these actions are inseparable.

Reply:
> What is the problem in this case?

It is readability issue. Let's ask for 3rd opinion.

@dkovalev1 , which option you consider is better?
```
ret->array[ret->count++] = pdl_node->xrelnode;
```
or
```
ret->array[ret->count] = pdl_node->xrelnode;
ret->count++;
```

Reply:
I don't think it is a big difference. The first one is more compact, I'd prefer it.